Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+import Foundation
+import class PusherPlatform.PPRequestOptions
+import enum PusherPlatform.HTTPMethod
+
+typealias SubscribeHandler = (VoidResult) -> Void
+
+protocol SubscriptionDelegate: AnyObject {
+    func subscription(_ subscription: Subscription, didReceiveEventWithJsonData jsonData: Data)
+    func subscription(_ subscription: Subscription, didReceiveError error: Error)
+}
+
+protocol Subscription {
+    func subscribe(completion: @escaping SubscribeHandler)
+    func unsubscribe()
+}
+
+class ConcreteSubscription: Subscription {
+    
+    enum State {
+        case notSubscribed
+        case subscribingStageOne(instanceWrapper: InstanceWrapper, completions: [SubscribeHandler])
```

Comment:
Could this be a bit more descriptive?

Reply:
Let me *try* and exlain why they're required...

`subscribingStageOne` -> `subscribeWithResume` has been invoked but has not returned. 
`subscribingStageTwo` -> `subscribeWithResume` has been invoked and has returned but `onOpen` has not yet fired.

`subscribingStageOne` and `subscribingStageTwo` are required because when calling `instanceWrapper.subscribeWithResume` its possible that the `onError` closure might fire BEFORE the method itself has returned.  In that instance we need to know within the `onError` closure that we were in the process of subscribing so we set the state to `subscribingStageOne` beforehand to indicate this state and it also holds reference to the completion handlers that are in need of invocation. 

If you have suggestion for a better name for the enum cases then I'd be happy to change.  Alternatively I can comment the code with the above.  Or if there's a better way to implement also happy for a suggestion.

Reply:
Since I took the time to write this I'm going to add it as a note/comment to the enum.